Move order latency control to the server

pchote opened the pull request:
This PR takes the order acks concept from #19588 and reworks it in a way that should be compatible with both current and new netcode systems, reducing the amount of parallel code and hopefully reducing the surface area for bugs.

The fundamental concept behind OpenRA's netcode is that clients generate packets that contain orders (attack X, move to Y) that are distributed to all clients, and then applied to the game state on some mutually agreed frame/tick. Getting these orders from each client to every other client takes a finite amount of time, so orders that are generated on net frame X (note that net frames are distinct from world ticks) need to be applied on net frame X + dX, where dX is what we call the order latency (typically 2-6 frames, or 6-18 ticks).

Our current plumbing uses a fixed latency for each game speed, which all clients manually add when serializing orders. This approach has worked well enough for many years, but can be done better. One of the big ideas from the netcode rework is to allow the server to dynamically vary the latency during gameplay, meaning that the common case can be much more responsive without making the game unplayable if network conditions degrade.

This PR takes the first few steps towards this by moving the calculation that applies the latency from the individual clients (before they send the orders) to the server (when the orders are routed to everyone else). This currently uses the same constant values as before - #19588 will replace this with its `OrderBuffer` system. It is not *quite* this simple, because the original client no longer knows when their orders should be applied. It would waste bandwidth if the server forwarded the players orders back to them, so it responds with just the frame number which the client combines with the original orders (which are stored in a new buffer) to give the same result to `OrderManager`. The server now also has to handle injecting empty frames at the start of the game, which is messy but (for now at least) an important part of the system.

I have taken this opportunity to clean up a bunch of other code and structural issues surrounding `OrderManager` and `Connection`. I strongly recommend reviewing by commit, with some extra context to help:

* ~~The first commit removes the one piece of gameplay code that depends on the order latency. Removing this now means we can remove all references to GameSpeed.OrderLatency from the client side. This is essentially replacing #9486 with a simpler (and hopefully temporary) workaround - the original issue was that `StructureProductionActiveDelay: 10` was not large enough for the fastest game speed, which has `OrderLatency: 6` ~ 18 ticks. I decided to use 25 because that is a nice round number (1 second) at the default game speed, and basically matches the current value at the fastest gamespeed. A future PR can hopefully fix this properly by explicitly accounting for just-placed structures and things in other production queues.~~
* ~~The second commit tries to implement a clear separation between `OrderManager` and `Connection`. `OrderManager` now only deals with engine-level concepts, and `Connection` (plus the new `OrdersPacket`) deals with converting those to and from raw byte streams. This should in theory be pure refactoring, with no changes to how/when things are processed.~~
* The third commit reduces some duplication in order serialization, split to its own commit because this does change how/when things are processed. Immediate orders are no longer forced to have their own individual packets, which has been redundant for many years now (afaik).
* The fourth commit splits the implementation of `EchoConnection` (used by the shellmap) and `NetworkConnection` (used by everything else). `EchoConnection` can now avoid serializating orders entirely. This may impact the order in which orders/immediate orders/sync packets are processed (but not the orders within each).
* The fifth commit finally implements the latency changes described at the start.

pchote committed:
Extract OrderIO.SerializeOrders helper.

Note: This changes immediate orders to no longer be
individually framed in their own packets.

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -31,109 +32,77 @@ public enum ConnectionState
-		void SendImmediate(IEnumerable<byte[]> orders);
-		void SendSync(int frame, byte[] syncData);
-		void Receive(Action<int, byte[]> packetFn);
+		void StartGame(OrderManager orderManager);
```
Remove the ordermanager parameter

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -621,6 +624,18 @@ byte[] CreateFrame(int client, int frame, byte[] data)
 
+		byte[] CreateAckFrame(int client, int frame)
+		{
+			using (var ms = new MemoryStream(13))
```
no need to dispose memorystream

pchote replied:
This (anti?)-pattern is used in quite a few places. Can we keep this here for consistency with `CreateFrame` above, and then have a followup PR that removes all of them?

teinarss commented on OpenRA.Game/Network/OrderIO.cs:
```diff
@@ -9,70 +9,148 @@
+			ms.WriteArray(BitConverter.GetBytes(frame));
+			foreach (var o in orders)
+				ms.WriteArray(o.Serialize());
+			return ms.ToArray();
```
GetBuffer?

pchote replied:
My understanding is that `GetBuffer` will return the full buffer, including potentially unallocated bytes that we don't want to send!

teinarss commented on OpenRA.Game/Network/OrderIO.cs:
```diff
@@ -9,70 +9,148 @@
+			ms.WriteArray(BitConverter.GetBytes(frame));
+			foreach (var o in orders)
+				ms.WriteArray(o.Serialize());
+			return ms.ToArray();
```
GetBuffer?

abcdefg30 commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -124,75 +120,71 @@ void CacheTextNotification(TextNotification notification)
+				UnitOrders.ProcessOrder(this, World, clientId, o);
+
+				// A mod switch or other event has pulled the ground from beneath us
+				if (disposed)
```
This will always be true for the shellmap, is it intentional that we just process one immediate order there? Or does this need a `&& World.Type != WorldType.Shellmap`?

pchote replied:
To be honest I just blindly copied over the existing checks. I will look into this, and may be able to drop some hacks.

pchote committed:
Fix shellmap OrderManager use after dispose.

pchote commented:
Added a new commit that fixes the OrderManager disposing mess.

pchote commented:
Also ping @jrb0001 for visibility.

jrb0001 commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -124,75 +120,55 @@ void CacheTextNotification(TextNotification notification)
+			if (pendingOrders.TryGetValue(clientId, out var queue))
+				queue.Enqueue((frame, orders));
+			else
+				Log.Write("debug", $"Received packet from disconnected client '{clientId}'");
```
This should probably get more visibility than just the debug log? It indicates a bug in the server implementation, and most likely a hard to find one.

pchote replied:
Agreed, and likewise when we receive an unknown packet type. I will have to think about the best way to expose this.

teinarss commented on OpenRA.Game/Network/OrderIO.cs:
```diff
@@ -9,70 +9,148 @@
+			foreach (var o in orders)
+				ms.WriteArray(o.Serialize());
+			return ms.ToArray();
 		}
```
Not used.

jrb0001 commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -281,17 +273,96 @@ protected override void Send(byte[] packet)
+
+			while (sentSync.TryDequeue(out var s))
+			{
+				orderManager.ReceiveSync(s.Frame, s.SyncHash, s.DefeatState);
```
This will desync if the server decides to push back orders from one client because they arrived late. Sync packets are small and not time critical so I think it would be better to always let them go through the server.

pchote replied:
I don't understand what you mean here, can you elaborate a bit futher?

There shouldn't be any desync when pushing back orders, provided that the server uses the same frame number in the ack packet to the sender as in the orders that it forwards to everyone else. Sync packet themselves shouldn't be pushed back by the server at all (they are a record of previous state, not future).

jrb0001 replied:
Ah you are right. Sync packets are attached to the last processed frame (can't change) while orders are attached to the future (can still be reordered).

jrb0001 commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -124,75 +120,55 @@ void CacheTextNotification(TextNotification notification)
-							Log.Write("debug", $"Received packet from disconnected client '{clientId}'");
-					}
-				});
+			pendingOrders.Remove(clientIndex);
```
I think this could lead to a desync if the server sends a disconnect right after some orders. The orders would get queued and then immediately discarded here. Depending on the network connection and the exact timing of the main loop, this might trigger on some clients while others only receive the disconnect on the next main loop iteration.

pchote replied:
I agree, and this is one of several things that is quite bogus in the disconnect handling code. However, this should be behaving exactly the same as it always has (if you can explain why it isn't then we will indeed need to fix it), so for the sake of unsticking progress on the netcode rework I would prefer that we defer any other changes to a future PR.

I'm thinking that the best solution wrt a future rewrite would be to have the server-side disconnect handling routed via the order buffer, which can then guarantee that all clients will see either a packet with orders in a given frame or a disconnect packet for the same frame but never both. The OrderManager code could then do something like queue a null into the pendingOrders queue as a deterministic disconnect marker that everyone would process at the same time.

jrb0001 replied:
It is a regression introduced by protocol version 13. Versions before that used a separate map with the frame number when the disconnect happened so it never discarded orders. My server implementation mitigates it by delaying the disconnect packet by one second, but that's very ugly.

It doesn't have to be in this PR, but It would be a good idea to fix it before the next playtest/release.

pchote replied:
Ok, thanks. I will look into whether there is a simple solution that can be added here.

pchote replied:
f1917251d37cf0bfa3e7f55b833085a6d822115a is what i'm currently thinking, however something must not be corrrect as this seems to trigger desyncs that I haven't yet debugged.

jrb0001 replied:
The implementation looks correct to me. It verifies that it is for the correct frame before applying it, so how could that desync? Maybe if the server sends different data to different clients, but that shouldn't happen, at least not in the disconnect case.

pchote replied:
I did a bit more debugging on this today, it looks like the desync might be related to incorrect Ack handling rather than disconnection. Starting a 3 player game and having players 1 and 2 issue orders while player 3 is stalling the game can repro the desync without anybody dropping.

pchote replied:
Ok, that was embarrasing. The problem was with the extra check that f1917251d37cf0bfa3e7f55b833085a6d822115a added to validate that acks were being sent by the server to match the disconnect handling - the server was writing these packets with the client's id, meaning that clients were ignoring their own orders...

pchote replied:
Fixed by #19640.

teinarss commented:
Needs a rebase

pchote commented:
Rebased.

penev92 commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -611,14 +614,22 @@ void ValidateClient(Connection newConn, string data)
 
 		byte[] CreateFrame(int client, int frame, byte[] data)
 		{
-			using (var ms = new MemoryStream(data.Length + 12))
```
I am maybe missing something obvious, but isn't `using` rather useful?

pchote replied:
https://github.com/OpenRA/OpenRA/pull/19632#discussion_r695533472

pchote commented:
Lets defer the disconnect fixes to a followup PR. The four commits here shouldn't change with that so theres no need to hold them up over it.

pchote commented:
Backported the (now correct!) client id checks from the disconnect commit into the ack commit. This should now be ready to go.

teinarss commented:
We could cleanup the passing  parameters to some of the functions, see https://github.com/teinarss/OpenRA/commit/61a2b15200169ac18e5ee06d48f32ef52b58e912

pchote commented:
Updated.

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -286,17 +274,98 @@ protected override void Send(byte[] packet)
+					orderManager.ReceiveDisconnect(disconnectClient);
+				else if (OrderIO.TryParseSync(p.Data, out var sync))
+					orderManager.ReceiveSync(sync);
+				else if (OrderIO.TryParseAck(p.FromClient, p.Data, out var ackFrame))
```
here you could pass `p`

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -123,41 +121,29 @@ void CacheTextNotification(TextNotification notification)
 		}
 
-		public void ReceiveSync(int frame, int syncHash, ulong defeatState)
+		public void ReceiveSync((int Frame, int SyncHash, ulong DefeatState) sync)
```
`in`

pchote commented:
Updated.

teinarss commented on OpenRA.Game/Network/OrderIO.cs:
```diff
@@ -91,43 +91,55 @@ public static bool TryParseDisconnect(byte[] packet, out int clientId)
+			return true;
+		}
+
+		public static bool TryParseAck(in (int FromClient, byte[] Data) packet, out int frame)
```
remove in?

pchote commented:
Whoops, fixed.

pchote committed:
Rework IConnection implementations:

* EchoConnection is now a trivial buffer that stores
  and repeats orders directly without serialization.

* NetworkConnection no longer subclasses EchoConnection,
  and now also caches local orders without serialization.

* Replay recording was moved to NetworkConnection
  (it is never used on EchoConnection).

abcdefg30 commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -252,18 +224,35 @@ void NetworkConnectionReceive()
+			var sync = (frame, syncHash, defeatState);
+			sentSync.Enqueue(sync);
+
+			// HACK: Defer sending this until after the orders (why?)
+			queuedSyncPackets.Add(OrderIO.SerializeSync(sync));
```
I don't understand this comment. Shouldn't it be at the point where we actually send the packets?

pchote replied:
This is just documenting that this is doing something unexpected for reasons that I don't understand. It is *probably* safe to just send the sync packets directly if you'd prefer.

abcdefg30 replied:
I see. Imo we can defer that to a follow-up if no more issues are found here that require bigger changes.

abcdefg30 reviewed (commented):
`FramesAhead` is still mentioned in OrderManager.cs#L208, we should probably slightly update that comment.

pchote committed:
Move order latency control to the server.

pchote commented:
Updated comments.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b81ebf5b1542d9b7e9ac497156fd21e6fce02519...88fa22994b1da27b6d30ac342eac72186d3940b9)

anvilvapre commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -31,117 +32,88 @@ public enum ConnectionState
+				// An immediate order may trigger a chain of actions that disposes the OrderManager and connection.
+				// Bail out to avoid potential problems from acting on disposed objects.
+				if (disposed)
+					break;
```
Shouldn't this be a `return` then as well. `orderManager` is used below.
